optimum/exporters/onnx/model_patcher.py

```diff
@@ -397,6 +397,219 @@ def __init__(
+
+    if labels is None:
+        final_labels = None
+
+    final_attention_mask = final_attention_mask.to(position_ids.dtype)
+
+    return final_embedding, final_attention_mask, final_labels, position_ids
+
+
+class LlavaModelPatcher(ModelPatcher):
+    def __enter__(self):
+        super().__enter__()
+        self.patch_ops()
+        setattr(
+            self._model,
+            "_merge_input_ids_with_image_features",
+            types.MethodType(patched_merge_input_ids_with_image_features, self._model),
+        )
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        super().__exit__(exc_type, exc_value, traceback)
+        self.restore_ops()
+        setattr(self._model, self.orig_forward_name, self.orig_forward)
+        setattr(
+            self._model,
+            "_merge_input_ids_with_image_features",
+            types.MethodType(self.original_merge_input_ids_with_image_features, self._model),
+        )
+
+    def __init__(
+        self,
+        config: "OnnxConfig",
+        model: Union["PreTrainedModel", "TFPreTrainedModel"],
+        model_kwargs: Optional[Dict[str, Any]] = None,
+    ):
+        super().__init__(config, model, model_kwargs)
+        self.original_merge_input_ids_with_image_features = self._model._merge_input_ids_with_image_features
+
+        def patched_forward(
+            input_ids=None,
+            pixel_values=None,
+            attention_mask=None,
+            position_ids=None,
+            past_key_values=None,
+            inputs_embeds=None,
+        ):
+            vision_feature_layer = model.config.vision_feature_layer
+            vision_feature_select_strategy = model.config.vision_feature_select_strategy
+
+            if config._behavior == "encoder":
+                inputs_embeds = model.get_input_embeddings()(input_ids)
+
+                image_outputs = model.vision_tower(pixel_values, output_hidden_states=True)
+                selected_image_feature = image_outputs.hidden_states[vision_feature_layer]
+
+                if vision_feature_select_strategy == "default":
+                    selected_image_feature = selected_image_feature[:, 1:]
+                elif vision_feature_select_strategy == "full":
+                    selected_image_feature = selected_image_feature
+                else:
+                    raise ValueError(f"Unexpected select feature strategy: {vision_feature_select_strategy}")
+
+                image_features = model.multi_modal_projector(selected_image_feature)
+                inputs_embeds, attention_mask, labels, position_ids = model._merge_input_ids_with_image_features(
+                    image_features, inputs_embeds, input_ids, attention_mask, None
+                )
+
+                result = {
+                    "inputs_embeds": inputs_embeds,
+                    "decoder_attention_mask": attention_mask,
+                    "position_ids": position_ids,
+                }
```

Comment:
I might not be understanding this 100%, but won't this be problematic for generation? We would need to re-pass the image features on every forward pass, which will merge the ids every time. This also means that we cannot embed a single text token (e.g., the one just generated).

Here's an example of a hand-crafted version of a tiny random LlavaForConditionalGeneration: https://huggingface.co/Xenova/tiny-random-LlavaForConditionalGeneration. There are 3 models exported:
- [embed_tokens.onnx](https://huggingface.co/Xenova/tiny-random-LlavaForConditionalGeneration/blob/main/onnx/embed_tokens.onnx) - just the token embedding layer
- [decoder_model_merged.onnx](https://huggingface.co/Xenova/tiny-random-LlavaForConditionalGeneration/blob/main/onnx/decoder_model_merged.onnx) - the causal LM
- [vision_encoder.onnx](https://huggingface.co/Xenova/tiny-random-LlavaForConditionalGeneration/blob/main/onnx/vision_encoder.onnx) - the vision encoder

I've got this working with Transformers.js (v3), where the concatenation of the token/vision patch embeddings are done in JavaScript.

Reply:
Hi @xenova, it should not be a problem for generation.

I generate the following three models:

1. `encoder_model.onnx` - Vision  tower + projection + merging
2. `decoder_model.onnx` - Language model only (The export is same as current decoder export in optimum)
3. `decoder_input_processor.onnx` - Decoder input generation when `past_key_values` is available. ([The elif part in the modeling code](https://github.com/huggingface/transformers/blob/v4.39.3/src/transformers/models/llava/modeling_llava.py#L421-L478))

The naming of models could possibly be updated.

This is how I use the models for inference: https://gist.github.com/mht-sharma/290f7bf9052e92023b4136c6fefd6717

ONNX Model: https://huggingface.co/mohitsha/llava-1.5-7b-hf/tree/main

In this version:  
1. I do all calculations as part of ONNX.
2. The embedding model is duplicated but is comparatively small.
3. If we want we could have additional 2 options for this part: 
           a. Create a separate `embed_model.onnx` and rest same. Now we have 4 ONNX models.
           b. Create a separate `embed_model.onnx` and do the `past_key_value` stage [attention_mask and position_ids processing](https://github.com/huggingface/transformers/blob/v4.39.3/src/transformers/models/llava/modeling_llava.py#L421-L478) as part of python code and remove `decoder_input_processor.onnx`

Let me know WDYT and if you have any suggestions.